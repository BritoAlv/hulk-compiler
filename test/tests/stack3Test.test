type Stack
{
    top = new Node(0);
    count = 0;

    push(n : number) : number => let pushedNode = new Node(n) in 
    {
        pushedNode.addPrevious(self.top);
        self.top.addNext(pushedNode);
        self.top := pushedNode;
        self.count := self.count + 1;
        n;
    };

    pop(n : number) : number => let popped = self.top.getValue() in 
    {
        self.top := self.top.getPrevious();
        self.count := self.count - 1;
        popped;
    };

    peek() : number => self.top.getValue();

    getCount() : number => self.count;
}

type Node(value : number)
{
    value = value;
    previous = self;
    next = self;
    addPrevious(previous : Node) : Node => self.previous := previous;
    addNext(next : Node) : Node => self.next := next;
    getPrevious() : Node => self.previous;
    getNext() : Node => self.next;
    print() : number => print(self.value);
    getValue() : number => self.value;
}

let stack = new Stack() in
{
    stack.push(4);
    print("Top:" @@ stack.peek());
    print("Count:" @@ stack.getCount());
    stack.push(3);
    stack.pop();
    stack.push(2);
    stack.push(5);
    stack.pop();
    stack.pop();
    print("Top:" @@ stack.peek());
    print("Count:" @@ stack.getCount());
    stack.push(10);
    print("Top:" @@ stack.peek());
    print("Count:" @@ stack.getCount());
};