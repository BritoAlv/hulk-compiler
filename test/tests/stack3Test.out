Top: 4
Count: 1
Top: 4
Count: 1
Top: 10
Count: 2

